lib/pbench/cli/server/repair.py

```diff
@@ -0,0 +1,320 @@
+from collections import defaultdict
+import copy
+import datetime
+from operator import and_
+from pathlib import Path
+from typing import Any, Optional
+
+import click
+from sqlalchemy import cast, column, or_, String, Subquery
+from sqlalchemy.orm import aliased
+from sqlalchemy.sql.expression import func
+
+from pbench.cli import pass_cli_context
+from pbench.cli.server import config_setup, Detail, Verify, Watch
+from pbench.cli.server.options import common_options
+from pbench.common.logger import get_pbench_logger
+from pbench.server import BadConfig, cache_manager
+from pbench.server.database.database import Database
+from pbench.server.database.models.audit import Audit
+from pbench.server.database.models.datasets import Dataset, Metadata
+from pbench.server.database.models.server_settings import get_retention_days
+from pbench.server.utils import UtcTimeHelper
+
+detailer: Optional[Detail] = None
+watcher: Optional[Watch] = None
+verifier: Optional[Verify] = None
+
+
+LIMIT = cache_manager.MAX_ERROR
+
+
+def repair_audit(kwargs):
+    """Repair certain audit record problems.
+
+    1. truncate extremely long audit event "attributes"
+
+    Args:
+        kwargs: the command options.
+    """
+
+    rows = (
+        Database.db_session.query(Audit)
+        .filter(func.length(cast(Audit.attributes, String)) >= LIMIT)
+        .order_by(Audit.timestamp)
+        .execution_options(stream_results=True)
+        .yield_per(2000)
+    )
+    count = 0
+    attributes_errors = 0
+    updated = 0
+    commit_error = None
+    for audit in rows:
+        count += 1
+        name = f"{audit.id}:{audit.status.name} {audit.name} {audit.object_name}"
+        # Deep copy works around a limitation in SQLAlchemy JSON proxy handling
+        a: dict[str, Any] = copy.deepcopy(audit.attributes)
+        if type(a) is not dict:
+            detailer.error(f"{name} attributes is type {type(a).__name__}")
+            attributes_errors += 1
+            continue
+        didit = False
+        for f in a:
+            value = a[f]
+            if type(value) is str and len(value) > LIMIT:
+                a[f] = value[:LIMIT]
+                detailer.message(f"{name} [{f}] truncated ({len(value)}) to {LIMIT}")
+                didit = True
+        if didit:
+            audit.attributes = a
+            updated += 1
+    if updated:
+        try:
+            Database.db_session.commit()
+        except Exception as e:
+            commit_error = str(e)
+    click.echo(f"{count} audit records had attributes too long")
+    click.echo(f"{updated} records were fixed")
```

Comment:
OK, so the trigger is the total size of the blob, but we only update it if one or more fields in the blob are over the limit...got it!

> our "target" tar errors could still trigger the query on a subsequent run, because {"error": "...."} may be longer than 5Kb even after truncating "..." to 5Kb

True, so if we perform a truncation, we should probably truncate it solidly _below_ the max, to leave space for the key, as well as a couple of braces, a couple of sets of quotes, a colon, and perhaps a few spaces.

Oh, and, did you want to add a trailing ellipsis or a leading `[TRUNCATED]` here, too?

Reply:
Did you forget to add the marker again?  😀

Once we hit the limit, I'm not really concerned about whether the last line is partial or not.  I _am_ concerned about providing enough context to diagnose the problem, so I would like to see the full message up to the limit, and I don't really care whether it's composed of multiple lines or not (e.g., the logger is probably going to remove/escape the line breaks, anyway...).  (So, if we have more than one line and the line which blows the limit is huge, I would like to see the first part of that line.)

And, regardless of boilerplate, there is nothing which says that the trigger for truncation _has_ to be related to the extent of the resulting truncation.  I think it would be good if a "successful" truncation on one pass didn't trigger another truncation (attempt) on another pass.  So, truncating below the max (instead of _to_ it) seems like a good idea.